### PR TITLE
Calculate SNR dB to Percentage based on Modulation

### PR DESCRIPTION
**Pull request by lars18th**
This patch incorporates a SNR value conversion from dB scale to percentage scale.
To do it, it uses a table of the correct dB values for each modulation. These values correspond to a 100% good signal.
The table includes at time only the common modulations. Other values could be included in the future.

### CODE REVIEW

File: src/dvb.c [hunk 1]
```diff
@@ -340,6 +340,234 @@ void copy_dvb_parameters(transponder *s, transponder *d) {
...
+                top = DVB_S2_PSK_8_FEC_8_9;
+                break;
+
+            default:
+                top = DVB_S2_PSK_8_OTHER;
+            }
+
+        default:
+            top = DVB_S2_OTHER;
+            LOG("get_db_snr_map -> Mudulation SNR scale mapping DVB-S2 not tested !!");
```
Comment (catalinii) [inline src/dvb.c:465]: Typo

Comment (lars18th) [reply src/dvb.c]: Fixed!

Comment (lars18th) [reply src/dvb.c]: Hi @catalinii ,

Improved with DVB-T2 support and with some range checks. Now the default behaviour is to use 80dB as the 100% SNR scale.

Please check it and merge.
Regards.

**Comment by catalinii** [comment]
@lars18th This PR added some coverity issues, can you fix them please: 

Hi,

Please find the latest report on new defect(s) introduced to minisatip2 found with Coverity Scan.

39 new defect(s) introduced to minisatip2 found with Coverity Scan.


New defect(s) Reported-by: Coverity Scan
Showing 20 of 39 defect(s)


** CID 367141:  Code maintainability issues  (UNUSED_VALUE)
/src/dvb.c: 497 in get_db_snr_map()


________________________________________________________________________________________________________
*** CID 367141:  Code maintainability issues  (UNUSED_VALUE)
/src/dvb.c: 497 in get_db_snr_map()
491                     top = DVB_T__QAM16_FEC_2_3;
492                     break;
493                 case FEC_3_4:
494                     top = DVB_T__QAM16_FEC_3_4;
495                     break;
496                 case FEC_5_6:
>>>     CID 367141:  Code maintainability issues  (UNUSED_VALUE)
>>>     Assigning value "DVB_T__QAM16_FEC_5_6" to "top" here, but that stored value is overwritten before it can be used.
497                     top = DVB_T__QAM16_FEC_5_6;
498                     break;
499                 case FEC_7_8:
500                     top = DVB_T__QAM16_FEC_7_8;
501                     break;
502     

** CID 367140:  Code maintainability issues  (UNUSED_VALUE)
/src/dvb.c: 464 in get_db_snr_map()


________________________________________________________________________________________________________
*** CID 367140:  Code maintainability issues  (UNUSED_VALUE)
/src/dvb.c: 464 in get_db_snr_map()
458             switch (tp->mtype) {
459     
460             case QPSK:
461             switch (tp->fec) {
462     
463                 case FEC_1_2:
>>>     CID 367140:  Code maintainability issues  (UNUSED_VALUE)
>>>     Assigning value "DVB_T__QPSK__FEC_1_2" to "top" here, but that stored value is overwritten before it can be used.
464                     top = DVB_T__QPSK__FEC_1_2;
465                     break;
466                 case FEC_2_3:
467                     top = DVB_T__QPSK__FEC_2_3;
468                     break;
469                 case FEC_3_4:

** CID 367139:  Code maintainability issues  (UNUSED_VALUE)
/src/dvb.c: 491 in get_db_snr_map()


________________________________________________________________________________________________________
*** CID 367139:  Code maintainability issues  (UNUSED_VALUE)
/src/dvb.c: 491 in get_db_snr_map()
485             switch (tp->fec) {
486     
487                 case FEC_1_2:
488                     top = DVB_T__QAM16_FEC_1_2;
489                     break;
490                 case FEC_2_3:
>>>     CID 367139:  Code maintainability issues  (UNUSED_VALUE)
>>>     Assigning value "DVB_T__QAM16_FEC_2_3" to "top" here, but that stored value is overwritten before it can be used.
491                     top = DVB_T__QAM16_FEC_2_3;
492                     break;
493                 case FEC_3_4:
494                     top = DVB_T__QAM16_FEC_3_4;
495                     break;
496                 case FEC_5_6:

** CID 367138:  Code maintainability issues  (UNUSED_VALUE)
/src/dvb.c: 577 in get_db_snr_map()


________________________________________________________________________________________________________
*** CID 367138:  Code maintainability issues  (UNUSED_VALUE)
/src/dvb.c: 577 in get_db_snr_map()
571                     top = DVB_T2_QAM16_FEC_1_2;
572                     break;
573                 case FEC_3_5:
574                     top = DVB_T2_QAM16_FEC_3_5;
575                     break;
576                 case FEC_2_3:
>>>     CID 367138:  Code maintainability issues  (UNUSED_VALUE)
>>>     Assigning value "DVB_T2_QAM16_FEC_2_3" to "top" here, but that stored value is overwritten before it can be used.
577                     top = DVB_T2_QAM16_FEC_2_3;
578                     break;
579                 case FEC_3_4:
580                     top = DVB_T2_QAM16_FEC_3_4;
581                     break;
582                 case FEC_4_5:

** CID 367137:  Code maintainability issues  (UNUSED_VALUE)
/src/dvb.c: 559 in get_db_snr_map()


________________________________________________________________________________________________________
*** CID 367137:  Code maintainability issues  (UNUSED_VALUE)
/src/dvb.c: 559 in get_db_snr_map()
553                     top = DVB_T2_QPSK__FEC_3_4;
554                     break;
555                 case FEC_4_5:
556                     top = DVB_T2_QPSK__FEC_4_5;
557                     break;
558                 case FEC_5_6:
>>>     CID 367137:  Code maintainability issues  (UNUSED_VALUE)
>>>     Assigning value "DVB_T2_QPSK__FEC_5_6" to "top" here, but that stored value is overwritten before it can be used.
559                     top = DVB_T2_QPSK__FEC_5_6;
560                     break;
561     
562                 default:
563                     top = DVB_T2_QPSK_OTHER;
564                     LOG("get_db_snr_map -> DVB-T2 modulation SNR scale not tested!!");

** CID 367136:  Code maintainability issues  (UNUSED_VALUE)
/src/dvb.c: 427 in get_db_snr_map()


________________________________________________________________________________________________________
*** CID 367136:  Code maintainability issues  (UNUSED_VALUE)
/src/dvb.c: 427 in get_db_snr_map()
421                     break;
422                 case FEC_9_10:
423                     top = DVB_S2_QPSK__FEC_9_10;
424                     break;
425     
426                 default:
>>>     CID 367136:  Code maintainability issues  (UNUSED_VALUE)
>>>     Assigning value "DVB_S2_OTHER" to "top" here, but that stored value is overwritten before it can be used.
427                     top = DVB_S2_OTHER;
428                     LOG("get_db_snr_map -> DVB-S2 modulation SNR scale not tested!!");
429                 }
430     
431             case PSK_8:
432             switch (tp->fec) {

** CID 367135:  Code maintainability issues  (UNUSED_VALUE)
/src/dvb.c: 411 in get_db_snr_map()


________________________________________________________________________________________________________
*** CID 367135:  Code maintainability issues  (UNUSED_VALUE)
/src/dvb.c: 411 in get_db_snr_map()
405             switch (tp->fec) {
406     
407                 case FEC_1_2:
408                     top = DVB_S2_QPSK__FEC_1_2;
409                     break;
410                 case FEC_2_3:
>>>     CID 367135:  Code maintainability issues  (UNUSED_VALUE)
>>>     Assigning value "DVB_S2_QPSK__FEC_2_3" to "top" here, but that stored value is overwritten before it can be used.
411                     top = DVB_S2_QPSK__FEC_2_3;
412                     break;
413                 case FEC_3_4:
414                     top = DVB_S2_QPSK__FEC_3_4;
415                     break;
416                 case FEC_5_6:

** CID 367134:  Code maintainability issues  (UNUSED_VALUE)
/src/dvb.c: 586 in get_db_snr_map()


________________________________________________________________________________________________________
*** CID 367134:  Code maintainability issues  (UNUSED_VALUE)
/src/dvb.c: 586 in get_db_snr_map()
580                     top = DVB_T2_QAM16_FEC_3_4;
581                     break;
582                 case FEC_4_5:
583                     top = DVB_T2_QAM16_FEC_4_5;
584                     break;
585                 case FEC_5_6:
>>>     CID 367134:  Code maintainability issues  (UNUSED_VALUE)
>>>     Assigning value "DVB_T2_QAM16_FEC_5_6" to "top" here, but that stored value is overwritten before it can be used.
586                     top = DVB_T2_QAM16_FEC_5_6;
587                     break;
588     
589                 default:
590                     top = DVB_T2_QAM16_OTHER;
591                     LOG("get_db_snr_map -> DVB-T2 modulation SNR scale not tested!!");

** CID 367133:  Code maintainability issues  (UNUSED_VALUE)
/src/dvb.c: 613 in get_db_snr_map()


________________________________________________________________________________________________________
*** CID 367133:  Code maintainability issues  (UNUSED_VALUE)
/src/dvb.c: 613 in get_db_snr_map()
607                     top = DVB_T2_QAM64_FEC_3_4;
608                     break;
609                 case FEC_4_5:
610                     top = DVB_T2_QAM64_FEC_4_5;
611                     break;
612                 case FEC_5_6:
>>>     CID 367133:  Code maintainability issues  (UNUSED_VALUE)
>>>     Assigning value "DVB_T2_QAM64_FEC_5_6" to "top" here, but that stored value is overwritten before it can be used.
613                     top = DVB_T2_QAM64_FEC_5_6;
614                     break;
615     
616                 default:
617                     top = DVB_T2_QAM64_OTHER;
618                     LOG("get_db_snr_map -> DVB-T2 modulation SNR scale not tested!!");

** CID 367132:  Code maintainability issues  (UNUSED_VALUE)
/src/dvb.c: 604 in get_db_snr_map()


________________________________________________________________________________________________________
*** CID 367132:  Code maintainability issues  (UNUSED_VALUE)
/src/dvb.c: 604 in get_db_snr_map()
598                     top = DVB_T2_QAM64_FEC_1_2;
599                     break;
600                 case FEC_3_5:
601                     top = DVB_T2_QAM64_FEC_3_5;
602                     break;
603                 case FEC_2_3:
>>>     CID 367132:  Code maintainability issues  (UNUSED_VALUE)
>>>     Assigning value "DVB_T2_QAM64_FEC_2_3" to "top" here, but that stored value is overwritten before it can be used.
604                     top = DVB_T2_QAM64_FEC_2_3;
605                     break;
606                 case FEC_3_4:
607                     top = DVB_T2_QAM64_FEC_3_4;
608                     break;
609                 case FEC_4_5:

** CID 367131:  Code maintainability issues  (UNUSED_VALUE)
/src/dvb.c: 417 in get_db_snr_map()


________________________________________________________________________________________________________
*** CID 367131:  Code maintainability issues  (UNUSED_VALUE)
/src/dvb.c: 417 in get_db_snr_map()
411                     top = DVB_S2_QPSK__FEC_2_3;
412                     break;
413                 case FEC_3_4:
414                     top = DVB_S2_QPSK__FEC_3_4;
415                     break;
416                 case FEC_5_6:
>>>     CID 367131:  Code maintainability issues  (UNUSED_VALUE)
>>>     Assigning value "DVB_S2_QPSK__FEC_5_6" to "top" here, but that stored value is overwritten before it can be used.
417                     top = DVB_S2_QPSK__FEC_5_6;
418                     break;
419                 case FEC_8_9:
420                     top = DVB_S2_QPSK__FEC_8_9;
421                     break;
422                 case FEC_9_10:

** CID 367130:  Code maintainability issues  (UNUSED_VALUE)
/src/dvb.c: 500 in get_db_snr_map()


________________________________________________________________________________________________________
*** CID 367130:  Code maintainability issues  (UNUSED_VALUE)
/src/dvb.c: 500 in get_db_snr_map()
494                     top = DVB_T__QAM16_FEC_3_4;
495                     break;
496                 case FEC_5_6:
497                     top = DVB_T__QAM16_FEC_5_6;
498                     break;
499                 case FEC_7_8:
>>>     CID 367130:  Code maintainability issues  (UNUSED_VALUE)
>>>     Assigning value "DVB_T__QAM16_FEC_7_8" to "top" here, but that stored value is overwritten before it can be used.
500                     top = DVB_T__QAM16_FEC_7_8;
501                     break;
502     
503                 default:
504                     top = DVB_T__QAM16_OTHER;
505                     LOG("get_db_snr_map -> DVB-T modulation SNR scale not tested!!");

** CID 367129:  Code maintainability issues  (UNUSED_VALUE)
/src/dvb.c: 420 in get_db_snr_map()


________________________________________________________________________________________________________
*** CID 367129:  Code maintainability issues  (UNUSED_VALUE)
/src/dvb.c: 420 in get_db_snr_map()
414                     top = DVB_S2_QPSK__FEC_3_4;
415                     break;
416                 case FEC_5_6:
417                     top = DVB_S2_QPSK__FEC_5_6;
418                     break;
419                 case FEC_8_9:
>>>     CID 367129:  Code maintainability issues  (UNUSED_VALUE)
>>>     Assigning value "DVB_S2_QPSK__FEC_8_9" to "top" here, but that stored value is overwritten before it can be used.
420                     top = DVB_S2_QPSK__FEC_8_9;
421                     break;
422                 case FEC_9_10:
423                     top = DVB_S2_QPSK__FEC_9_10;
424                     break;
425     

** CID 367128:  Code maintainability issues  (UNUSED_VALUE)
/src/dvb.c: 467 in get_db_snr_map()


________________________________________________________________________________________________________
*** CID 367128:  Code maintainability issues  (UNUSED_VALUE)
/src/dvb.c: 467 in get_db_snr_map()
461             switch (tp->fec) {
462     
463                 case FEC_1_2:
464                     top = DVB_T__QPSK__FEC_1_2;
465                     break;
466                 case FEC_2_3:
>>>     CID 367128:  Code maintainability issues  (UNUSED_VALUE)
>>>     Assigning value "DVB_T__QPSK__FEC_2_3" to "top" here, but that stored value is overwritten before it can be used.
467                     top = DVB_T__QPSK__FEC_2_3;
468                     break;
469                 case FEC_3_4:
470                     top = DVB_T__QPSK__FEC_3_4;
471                     break;
472                 case FEC_5_6:

** CID 367127:  Code maintainability issues  (UNUSED_VALUE)
/src/dvb.c: 414 in get_db_snr_map()


________________________________________________________________________________________________________
*** CID 367127:  Code maintainability issues  (UNUSED_VALUE)
/src/dvb.c: 414 in get_db_snr_map()
408                     top = DVB_S2_QPSK__FEC_1_2;
409                     break;
410                 case FEC_2_3:
411                     top = DVB_S2_QPSK__FEC_2_3;
412                     break;
413                 case FEC_3_4:
>>>     CID 367127:  Code maintainability issues  (UNUSED_VALUE)
>>>     Assigning value "DVB_S2_QPSK__FEC_3_4" to "top" here, but that stored value is overwritten before it can be used.
414                     top = DVB_S2_QPSK__FEC_3_4;
415                     break;
416                 case FEC_5_6:
417                     top = DVB_S2_QPSK__FEC_5_6;
418                     break;
419                 case FEC_8_9:

** CID 367126:  Code maintainability issues  (UNUSED_VALUE)
/src/dvb.c: 580 in get_db_snr_map()


________________________________________________________________________________________________________
*** CID 367126:  Code maintainability issues  (UNUSED_VALUE)
/src/dvb.c: 580 in get_db_snr_map()
574                     top = DVB_T2_QAM16_FEC_3_5;
575                     break;
576                 case FEC_2_3:
577                     top = DVB_T2_QAM16_FEC_2_3;
578                     break;
579                 case FEC_3_4:
>>>     CID 367126:  Code maintainability issues  (UNUSED_VALUE)
>>>     Assigning value "DVB_T2_QAM16_FEC_3_4" to "top" here, but that stored value is overwritten before it can be used.
580                     top = DVB_T2_QAM16_FEC_3_4;
581                     break;
582                 case FEC_4_5:
583                     top = DVB_T2_QAM16_FEC_4_5;
584                     break;
585                 case FEC_5_6:

** CID 367125:  Code maintainability issues  (UNUSED_VALUE)
/src/dvb.c: 598 in get_db_snr_map()


________________________________________________________________________________________________________
*** CID 367125:  Code maintainability issues  (UNUSED_VALUE)
/src/dvb.c: 598 in get_db_snr_map()
592                 }
593     
594             case QAM_64:
595             switch (tp->fec) {
596     
597                 case FEC_1_2:
>>>     CID 367125:  Code maintainability issues  (UNUSED_VALUE)
>>>     Assigning value "DVB_T2_QAM64_FEC_1_2" to "top" here, but that stored value is overwritten before it can be used.
598                     top = DVB_T2_QAM64_FEC_1_2;
599                     break;
600                 case FEC_3_5:
601                     top = DVB_T2_QAM64_FEC_3_5;
602                     break;
603                 case FEC_2_3:

** CID 367124:  Code maintainability issues  (UNUSED_VALUE)
/src/dvb.c: 547 in get_db_snr_map()


________________________________________________________________________________________________________
*** CID 367124:  Code maintainability issues  (UNUSED_VALUE)
/src/dvb.c: 547 in get_db_snr_map()
541             switch (tp->fec) {
542     
543                 case FEC_1_2:
544                     top = DVB_T2_QPSK__FEC_1_2;
545                     break;
546                 case FEC_3_5:
>>>     CID 367124:  Code maintainability issues  (UNUSED_VALUE)
>>>     Assigning value "DVB_T2_QPSK__FEC_3_5" to "top" here, but that stored value is overwritten before it can be used.
547                     top = DVB_T2_QPSK__FEC_3_5;
548                     break;
549                 case FEC_2_3:
550                     top = DVB_T2_QPSK__FEC_2_3;
551                     break;
552                 case FEC_3_4:

** CID 367123:  Code maintainability issues  (UNUSED_VALUE)
/src/dvb.c: 473 in get_db_snr_map()


________________________________________________________________________________________________________
*** CID 367123:  Code maintainability issues  (UNUSED_VALUE)
/src/dvb.c: 473 in get_db_snr_map()
467                     top = DVB_T__QPSK__FEC_2_3;
468                     break;
469                 case FEC_3_4:
470                     top = DVB_T__QPSK__FEC_3_4;
471                     break;
472                 case FEC_5_6:
>>>     CID 367123:  Code maintainability issues  (UNUSED_VALUE)
>>>     Assigning value "DVB_T__QPSK__FEC_5_6" to "top" here, but that stored value is overwritten before it can be used.
473                     top = DVB_T__QPSK__FEC_5_6;
474                     break;
475                 case FEC_7_8:
476                     top = DVB_T__QPSK__FEC_7_8;
477                     break;
478     

** CID 367122:  Code maintainability issues  (UNUSED_VALUE)
/src/dvb.c: 607 in get_db_snr_map()


________________________________________________________________________________________________________
*** CID 367122:  Code maintainability issues  (UNUSED_VALUE)
/src/dvb.c: 607 in get_db_snr_map()
601                     top = DVB_T2_QAM64_FEC_3_5;
602                     break;
603                 case FEC_2_3:
604                     top = DVB_T2_QAM64_FEC_2_3;
605                     break;
606                 case FEC_3_4:
>>>     CID 367122:  Code maintainability issues  (UNUSED_VALUE)
>>>     Assigning value "DVB_T2_QAM64_FEC_3_4" to "top" here, but that stored value is overwritten before it can be used.
607                     top = DVB_T2_QAM64_FEC_3_4;
608                     break;
609                 case FEC_4_5:
610                     top = DVB_T2_QAM64_FEC_4_5;
611                     break;
612                 case FEC_5_6:

**Comment by lars18th** [comment]
Hi @catalinii ,

I don't understand these coverity warnings. The code works without troubles, and the "top" is used _everytime_ in the line `db_map = top / 10.0;`. So from my point of view, the warning doesn't have sense.

**Comment by jo-heremans** [comment]
Some case blocks are not ended by break.  Add break on line 430, so the value of line 411 is not overwritten on line 435

**Comment by lars18th** [comment]
> Some case blocks are not ended by break. Add break on line 430, so the value of line 411 is not overwritten on line 435

You're right. Thank you to point it. I'll solve asap. 👍

**Comment by lars18th** [comment]
Fixed in #913 😉

**Pull request merged**